File: 2023/day_2/main.go
package main

import (
    "fmt"
    "os"
    "strings"
    "strconv"
)

func part_1() {
    content, err := os.ReadFile("input.txt")
    if err != nil {
        panic(err)
    }
    lines := strings.Split(string(content), "\n")
    lines = lines[:len(lines)-1]

    sum := 0
    for game, line := range lines {
        content := strings.Split(line, ":")[1]
        sets := strings.Split(content, ";")
        
        var balls []string
        for _, set := range sets {
            balls = append(balls, strings.Split(set, ",")...)
        }
        for i, ball := range balls {
            balls[i] = strings.TrimSpace(ball)
        }
        
        possible := true
        for _, ball := range balls {
            number, err := strconv.Atoi(strings.Split(ball, " ")[0])
            if err != nil {
                panic(err)
            }
            color := strings.Split(ball, " ")[1]
            
            if color == "red" && number > 12 {
                possible = false
                break
            } else if color == "green" && number > 13 {
                possible = false
                break
            } else if color == "blue" && number > 14 {
                possible = false
                break
            }
        }

        if possible {
            sum += game + 1
        }
    }
    fmt.Println(sum)
}

func part_2() {
    content, err := os.ReadFile("input.txt")
    if err != nil {
        panic(err)
    }
    lines := strings.Split(string(content), "\n")
    lines = lines[:len(lines)-1]

    sum := 0
    for _, line := range lines {
        content := strings.Split(line, ":")[1]
        sets := strings.Split(content, ";")
        
        var balls []string
        for _, set := range sets {
            balls = append(balls, strings.Split(set, ",")...)
        }
        for i, ball := range balls {
            balls[i] = strings.TrimSpace(ball)
        }
        
        maxRed := 0
        maxGreen := 0
        maxBlue := 0
        for _, ball := range balls {
            number, err := strconv.Atoi(strings.Split(ball, " ")[0])
            if err != nil {
                panic(err)
            }
            color := strings.Split(ball, " ")[1]

            if color == "red" && number > maxRed {
                maxRed = number
            } else if color == "green" && number > maxGreen {
                maxGreen = number
            } else if color == "blue" && number > maxBlue {
                maxBlue = number
            }
        }
        sum += maxRed * maxGreen * maxBlue
    }
    fmt.Println(sum)
}

func main() {
    part_1()
    part_2()
}

File: 2023/day_2/main.py


def part_1() -> None:
    with open("input.txt") as f:
        lines: list[str] = f.read().splitlines()

    sum: int = 0
    for game, line in enumerate(lines):
        content: str = line.split(":")[1]
        sets: list[str] = content.split(";")
        sets = [set_.strip() for set_ in sets]

        balls: list[str] = []
        for set_ in sets:
            balls += set_.split(",")
        balls = [ball.strip() for ball in balls]

        possible: bool = True
        for ball in balls:
            color: str = ball.split(" ")[1]
            number: int = int(ball.split(" ")[0])

            if color == "red" and number > 12:
                possible = False
                break
            elif color == "green" and number > 13:
                possible = False
                break
            elif color == "blue" and number > 14:
                possible = False
                break

        if possible:
            sum += game + 1

    print(sum)


def part_2() -> None:
    with open("input.txt") as f:
        lines: list[str] = f.read().splitlines()

    sum: int = 0
    for line in lines:
        content: str = line.split(":")[1]
        sets: list[str] = content.split(";")
        sets = [set_.strip() for set_ in sets]

        balls: list[str] = []
        for set_ in sets:
            balls += set_.split(",")
        balls = [ball.strip() for ball in balls]

        max_red: int = 0
        max_green: int = 0
        max_blue: int = 0
        for ball in balls:
            color: str = ball.split(" ")[1]
            number: int = int(ball.split(" ")[0])

            if color == "red" and number > max_red:
                max_red = number
            elif color == "green" and number > max_green:
                max_green = number
            elif color == "blue" and number > max_blue:
                max_blue = number

        sum += max_red*max_green*max_blue

    print(sum)


if __name__ == "__main__":
    part_1()
    part_2()
